### PR TITLE
fix: skip compressed assistant tool calls

### LINKED ISSUE #1482: AutoContextMemory导致工具重复调用，直到maxIters才结束，触发压缩后有概率出现这个问题
以下为AI分析结果：
两个高风险点：AutoContextHook 会重写 PreReasoningEvent 的输入消息；AutoContextMemory 多处注释提到“保留 ToolUseBlock 结构”。如果压缩后的上下文记忆仍带历史 ToolUseBlock， ReAct 循环很可能把它当作新的待执行工具请求。

更具体的线索：大消息摘要路径会把 ToolUseBlock 原样放回压缩后的 message，注释还明确说“preserve ToolUseBlock structure”。这对 API 消息格式也许是为了配对工具结果，但对 ReAct的“是否有工具调用需要执行”的判定很危险。

工具执行后会立刻进入下一次 reasoning，而 AutoContextHook 在这次 reasoning 前调用压缩。也就是说压缩可能发生在“工具结果刚写入 memory、模型还没基于这个 tool_result生成最终回答”之间；如果这时把当前轮 tool_use/tool_result 压成普通摘要，模型很容易把它当作历史叙述而不是刚返回的工具结果，从而再次请求同一个工具。

问题根因在 AutoContextHook 的 PreReasoningEvent：工具执行结果刚写入 memory 后，下一轮 reasoning 前会触发压缩。原实现直接走 compressIfNeeded()，可能把“当前轮尚未被模型消费的tool_use/tool_result”压成普通摘要，模型看不到结构化工具结果，就容易再次调用同一个工具。

### PR DESCRIPTION
**Pull request by guslegend0510**
### 背景
在 AutoContext 压缩场景下，assistant 消息会带有 `_compress_meta` 标记。原有逻辑在提取最近工具调用时，可能把这些压缩后的消息当成新的工具调用上下文，导致后续轮次恢复出的工具调用不正确，甚至出现重复执行风险。

### 修改内容
- 在 `MessageUtils.extractRecentToolCalls()` 中跳过带 `_compress_meta` 的 assistant 消息。
- 避免压缩消息污染“最近工具调用”的提取结果。
- 保持普通 assistant 消息和原有工具调用处理逻辑不变。

### 验证
- `mvn -pl agentscope-core -am -Dtest=MessageUtilsTest test`

Fixes #1482

### CODE REVIEW

**Comment by LearningGp** [comment]
image

**Comment by guslegend0510** [comment]
@LearningGp 麻烦审核一下，谢谢

File: agentscope-core/src/main/java/io/agentscope/core/util/MessageUtils.java [hunk 1]
```diff
@@ -63,4 +69,9 @@ public static List<ToolUseBlock> extractRecentToolCalls(List<Msg> messages, Stri
 
         return List.of();
     }
+
+    private static boolean isCompressedMessage(Msg msg) {
+        Map<String, Object> metadata = msg.getMetadata();
```
Comment (AgentScopeJavaBot) [inline agentscope-core/src/main/java/io/agentscope/core/util/MessageUtils.java:82]: [nit] Consider adding a defensive null check on `metadata` to match the existing pattern within `Msg` itself (e.g., `Msg.getChatUsage()` at line 564 checks `if (metadata == null)` before accessing it). While the `Msg` constructor currently guarantees a non-null `HashMap`, the `getMetadata()` Javadoc states it *can* return null, and following the established defensive pattern is safer:

```java
private static boolean isCompressedMessage(Msg msg) {
    Map<String, Object> metadata = msg.getMetadata();
    return metadata != null && metadata.get(COMPRESS_META_KEY) instanceof Map;
}
```

File: agentscope-core/src/test/java/io/agentscope/core/util/MessageUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+class MessageUtilsTest {
+
+    @Test
+    @DisplayName("Should skip compressed assistant messages when extracting recent tool calls")
+    void testExtractRecentToolCallsSkipsCompressedAssistantMessages() {
+        List<Msg> messages = new ArrayList<>();
+        messages.add(createAssistantToolUseMessage("real-tool", "real-call"));
+        messages.add(createCompressedAssistantToolUseMessage("compressed-tool", "compressed-call"));
+
+        List<ToolUseBlock> toolCalls = MessageUtils.extractRecentToolCalls(messages, "assistant");
```
Comment (AgentScopeJavaBot) [inline agentscope-core/src/test/java/io/agentscope/core/util/MessageUtilsTest.java:43]: [nit] Test coverage is good for the main scenarios. Consider adding one more edge case: when **all** assistant messages in the list are compressed, the method should return an empty list. This validates that the `continue` loop terminates correctly without falling through to stale tool calls.

**Review by AgentScopeJavaBot — APPROVED** [review]
🤖 AI Review

This PR fixes a bug where `MessageUtils.extractRecentToolCalls()` could pick up tool calls from compressed assistant messages (those carrying `_compress_meta` metadata), leading to incorrect or duplicate tool call extraction in the AutoContext compression scenario. The fix is clean and well-scoped: it adds a `continue` guard to skip compressed messages when scanning backwards, and includes a new test class covering the main cases. The logic is correct and aligns with the issue description.

File: agentscope-core/src/main/java/io/agentscope/core/util/MessageUtils.java (hunk 1 above)
Comment (guslegend0510) [reply agentscope-core/src/main/java/io/agentscope/core/util/MessageUtils.java:82]: 在 isCompressedMessage 里补了 metadata 的空判，避免这里隐式依赖非空返回。

File: agentscope-core/src/test/java/io/agentscope/core/util/MessageUtilsTest.java (hunk 2 above)
Comment (guslegend0510) [reply agentscope-core/src/test/java/io/agentscope/core/util/MessageUtilsTest.java:43]: 收紧了 extractRecentToolCalls 的回溯范围，现在只会在最近一段连续的 assistant 消息里查找，避免跳过压缩后的 continue 消息后又回退到更早的旧 tool call。
